language/move-lang/src/typing/core.rs

```diff
@@ -897,103 +905,110 @@ pub fn solve_constraints(context: &mut Context) {
-                Kind_::RESOURCE_CONSTRAINT
-            );
-            context.error(vec![
-                (loc, "Constraint not satisfied.".into()),
-                (bloc, cmsg),
-                (
-                    b_kind.loc,
-                    "The type's constraint information was determined here".into(),
-                ),
-                (
-                    k.loc,
-                    format!("'{}' constraint declared here", Kind_::RESOURCE_CONSTRAINT),
-                ),
-            ])
+fn solve_ability_constraint(
+    context: &mut Context,
+    loc: Loc,
+    given_msg_opt: Option<String>,
+    ty: Type,
+    constraints: AbilitySet,
+) {
+    let ty = unfold_type(&context.subst, ty);
+    let ty_loc = ty.loc;
+    let ty_abilities = infer_abilities(context, &context.subst, ty.clone());
+
+    let (declared_loc_opt, declared_abilities, ty_args) = debug_abilities_info(context, &ty);
+    for constraint in constraints {
+        if ty_abilities.has_ability(&constraint).is_some() {
+            continue;
         }
 
-        // all </: affine
-        // linear </: affine
-        (bk @ K::Unknown, K::Affine) | (bk @ K::Resource, K::Affine) => {
-            let resource_msg = match bk {
-                K::Copyable | K::Affine => panic!("ICE covered above"),
-                K::Resource => "resource ",
-                K::Unknown => "",
-            };
-            let ty_str = error_format(&b, &context.subst);
-            let cmsg = format!(
-                "The {}type {} does not satisfy the constraint '{}'",
-                resource_msg,
-                ty_str,
-                Kind_::VALUE_CONSTRAINT
-            );
-            context.error(vec![
-                (loc, "Constraint not satisfied.".into()),
-                (bloc, cmsg),
-                (
-                    b_kind.loc,
-                    "The type's constraint information was determined here".into(),
-                ),
-                (
-                    k.loc,
-                    format!("'{}' constraint declared here", Kind_::VALUE_CONSTRAINT),
+        let ty_str = error_format(&ty, &context.subst);
+        let constraint_msg = match &given_msg_opt {
+            Some(s) => s.clone(),
+            None => format!("'{}' constraint not satisifed", constraint),
+        };
+        let ty_msg = format!(
+            "The type {} does not have the ability '{}'",
+            ty_str, constraint
+        );
+        let mut error = vec![(loc, constraint_msg), (ty_loc, ty_msg)];
+
+        match (
+            declared_loc_opt,
+            declared_abilities.has_ability(&constraint),
+        ) {
+            // primtive or vector or "tuple" does not have ability
+            (None, _) => (),
+
+            // Type was not given the ability
+            (Some(dloc), None) => error.push((
+                dloc,
+                format!(
+                    "To satisify the constraint, the '{}' ability would need to be added here",
```

Comment:
Typo: "satisify"
This one shows up in an error message, so please be sure to fix it.